linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/engine/DefaultEngineCreateService.java

```diff
@@ -0,0 +1,429 @@
+                engineNode.getServiceInstance(), resourceTicketId, errorInfo.getKey()));
+      }
+      throw new AMErrorException(
+          AMConstant.EM_ERROR_CODE,
+          String.format(
+              "%s ticketID: %s Failed to initialize engine, reason: %s",
+              engineNode.getServiceInstance(), resourceTicketId, errorInfo.getKey()));
+    }
+    return NodeStatus.isAvailable(engineNodeInfo.getNodeStatus());
+  }
+
+  private Pair<String, Optional<Boolean>> getStartErrorInfo(String msg) {
+    if (StringUtils.isNotBlank(msg)) {
+      JsonNode jsonNode = null;
+      try {
+        jsonNode = BDPJettyServerHelper.jacksonJson().readTree(msg);
+      } catch (JsonProcessingException e) {
+        logger.warn("getStartErrorInfo readTree failed msg: {}", msg, e);
+      }
+      if (jsonNode != null && jsonNode.has(AMConstant.START_REASON)) {
+        String startReason = jsonNode.get(AMConstant.START_REASON).asText();
+        if (jsonNode.has(AMConstant.EC_CAN_RETRY)) {
+          return Pair.of(startReason, Optional.of(true));
+        } else {
+          return Pair.of(startReason, Optional.empty());
+        }
+      }
+    }
+    return Pair.of(null, Optional.empty());
+  }
+
+  /**
+   * Need to ensure that the newly created amount ec is available before returning
+   *
+   * @param engineNode
+   * @param resourceTicketId
+   * @param timeout
+   * @return
+   */
+  public EngineNode ensureECAvailable(
+      EngineNode engineNode, String resourceTicketId, long timeout) {
+    try {
+      logger.info(
+          String.format(
+              "Start to wait engineConn(%s) to be available, but only %s left.",
+              engineNode, ByteTimeUtils.msDurationToString(timeout)));
+      LinkisUtils.waitUntil(
+          () -> ensuresIdle(engineNode, resourceTicketId), Duration.ofMillis(timeout));
+    } catch (TimeoutException e) {
+      logger.info(
+          String.format(
+              "Waiting for engineNode:%s(%s) initialization TimeoutException , now stop it.",
+              engineNode, resourceTicketId));
+      EngineStopRequest stopEngineRequest =
+          new EngineStopRequest(engineNode.getServiceInstance(), ManagerUtils.getAdminUser());
+      engineStopService.asyncStopEngine(stopEngineRequest);
+      throw new LinkisRetryException(
+          AMConstant.ENGINE_ERROR_CODE,
+          String.format(
+              "Waiting for engineNode:%s(%s) initialization TimeoutException, already waiting %d ms",
+              engineNode, resourceTicketId, timeout));
+    } catch (Throwable t) {
+      logger.info(
+          String.format(
+              "Waiting for %s(%s) initialization failure , now stop it.",
+              engineNode, resourceTicketId));
+      EngineStopRequest stopEngineRequest =
+          new EngineStopRequest(engineNode.getServiceInstance(), ManagerUtils.getAdminUser());
+      engineStopService.asyncStopEngine(stopEngineRequest);
+      throw new RuntimeException(t);
```

Comment:
Should LinkisRuntimeException

Reply:
used LinkisRuntimie instead of RuntimeException, thanks.